llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -16878,6 +16870,16 @@ SDValue DAGCombiner::visitFMULForFMADistributiveCombine(SDNode *N) {
   if (!hasNoInfs(Options, FAdd))
     return SDValue();
 
+  // Returns true if floating point contraction is allowed on the FMUL-SDValue
+  // `N`
+  auto isContractableFMUL = [this](const TargetOptions &Options, SDValue N) {
+    assert(N.getOpcode() == ISD::FMUL);
+
+    return Options.AllowFPOpFusion == FPOpFusion::Fast ||
+           DAG.getMachineFunction().hasUnsafeFPMath() ||
```

Comment:
This is one of the most egregious uses of the unsafe-fp-math. It is interpreted inconsistently and confusingly. In this case, we already have the global fusion option and the contract flag. 

In a separate PR, can you drop the usage of unsafe fp math here?